### ISSUE TITLE
New lint: `duplicate_match_guards`

### ISSUE THREAD

**Issue by ada4a** [rust-lang/rust-clippy#14971]
### What it does

Checks for the same condition being checked in a match guard and in the match body

### Advantage

- Remove a likely typo or a copy and paste error

### Drawbacks

- False negatives: if the condition is an impure function, it could've been called twice on purpose for its side effects
- removing a match guard might change the matching logic (so it might make sense to only suggest removing the inner `if`)

### Example

```rust
match n {
    0 if a > b => {
        if a > b {
            return;
        }
    }
    _ => {}
}
```

Could be written as:

```rust
match n {
    0 if a > b => {
        return;
    }
    _ => {}
}
```

**Comment by ada4a** [comment]
I think the logic will be largely similar to [`eq_op`](https://rust-lang.github.io/rust-clippy/master/index.html#eq_op)

But there are some additional considerations:
1) the match body should be one of:
    - `ExprKind::Block` with no `stmts` and the `if` as the `trailing_expr`
    - `ExprKind::If`
2) the inner `if` may not have an `else`

**Comment by ada4a** [comment]
I might work on this myself, but I should probably finish https://github.com/rust-lang/rust-clippy/pull/14895 first 😅
